azurerm/resource_arm_redis_cache.go

```diff
@@ -660,7 +673,16 @@ func expandRedisConfiguration(d *schema.ResourceData) map[string]*string {
 		output["aof-storage-connection-string-1"] = utils.String(v.(string))
 	}
 
-	return output
+	// Redis authentication can only be disabled if it is launched inside a VNET.
+	if v, ok := d.GetOkExists("redis_configuration.0.enable_authentication"); ok {
+		if _, isPrivate := d.GetOk("subnet_id"); !isPrivate {
```

Comment:
checking the tests - we also need to verify that authentication's disabled here:

```suggestion
		if _, isPrivate := d.GetOk("subnet_id"); !isPrivate && !v {
```

Reply:
It was something similar but this caused the acceptance test cases to fail as @katbyte had commented - this is because Azure does not accept `authnotrequired` to be passed at all if the cache is public. That's the reason there's a blanket check whether if its public or not

Reply:
got it - in which case I think we want this set a level higher? e.g.

```
	if v, ok := d.GetOkExists("redis_configuration.0.enable_authentication"); ok {
          authenticationEnabled := v.(bool)
          if !authenticationEnabled {
             // assertion and assignment
          }
```

WDYT?

(p.s. sorry for messing up the commits, unfortunately I can't push to your master branch without breaking the PR)

Reply:
Ok, so I think there's some communication gap here. So there are two ways to look at the condition:
i) Don't allow `enable_authentication` to be set for public caches.
ii) Allow `enable_authentication` to be set for public caches but only accept `true` as the value.

The changes in the PR assumed we are going with i). Do we want option ii)? If so I can make the necessary changes for this.

Reply:
Made the changes for this
```
		authEnabled := v.(bool)
		_, isPrivate := d.GetOk("subnet_id")

		// Redis authentication can only be disabled if it is launched inside a VNET.
		if !isPrivate {
			if !authEnabled {
				return nil, fmt.Errorf("Cannot set `enable_authentication` to `false` when `subnet_id` is not set")
			}
		} else {
			value := isAuthNotRequiredAsString(authEnabled)
			output["authnotrequired"] = utils.String(value)
		}
	}
```

Is this fine ??

Reply:
@code-haven looks good - thanks for pushing that 👍